net: 6LoCAN implementation

alexanderwachter opened the pull request:
This is an implementation of 6LoCAN, a 6Lo adaption layer for Controller Area Networks.
6LoCAN is not yet standardized and experimental.
With 6LoCAN you can use CAN as data link layer for IPv6. Linklayer addresses are assigned randomly or defined in Kconfig. To avoid collisions  in link layer addresses, a ll duplicate address detection is performed. IPv6 packets are transmitted using ISO-TP.

jukkar commented on drivers/can/Kconfig.net:
```diff
@@ -0,0 +1,38 @@
+
+config CAN_NET
+	bool "Enable 6loCAN network interface [Experimental]"
+	default n
```
"default n" strings can be omitted because by default the boolean variables are false

jukkar commented on drivers/can/Kconfig.net:
```diff
@@ -0,0 +1,38 @@
+#
+
+config CAN_NET
+	bool "Enable 6loCAN network interface [Experimental]"
```
very minor nitpick, please write "EXPERIMENTAL" (all capital) like in other parts of the system.

jukkar commented on drivers/can/can_net.c:
```diff
@@ -0,0 +1,255 @@
+	net_pkt_lladdr_src(pkt)->len = sizeof(struct net_can_lladdr);
+	net_pkt_lladdr_src(pkt)->type = NET_LINK_CANBUS;
+	net_buf_add_be16(buf, get_can_lladdr_src(frame));
+	net_buf_pull(buf, sizeof(u16_t));
```
Why net_buf_pull() here, the data before the data pointer is now inaccesible?

alexanderwachter replied:
This bytes are the lladdr now. they are referenced by the pkt. The rest of the data is the data in the frame then.

jukkar replied:
> This bytes are the lladdr now. they are referenced by the pkt. The rest of the data is the data in the frame then.

Yes, I understood that. It is just that all of our APIs kind of do not see that any more because the buf->data pointer is moved forward. Not sure if this is an issue thou.

alexanderwachter replied:
I did this because net_pkt_cursor_init sets the cursor to the actual data then instead of the  link layer address. Otherwise i have to skip the first 4 bytes everywhere else in the code.

alexanderwachter replied:
@jukkar This indeed needs some rework. @tbursztyka pointed me to a better solution.

alexanderwachter replied:
@jukkar I tried to remove it here but it seems to make difficulties in canbus.c.
The problem is that there are several situations where the first bytes would have to be skipped. IMO it's better to just pull them here.

jukkar replied:
> IMO it's better to just pull them here.

Ok.

jukkar commented on drivers/can/can_net.c:
```diff
@@ -0,0 +1,255 @@
+	return can_send(ctx->can_dev, frame, timeout, cb, cb_arg);
+}
+
+static void net_can_rcv(struct zcan_frame *frame, void *arg)
```
please use _recv instead of _rcv (for consistency with other part of networking code)

jukkar commented on drivers/can/can_net.c:
```diff
@@ -0,0 +1,255 @@
+
+	return 0;
+}
+
```
remove extra empty space

jukkar commented on drivers/can/Kconfig.net:
```diff
@@ -0,0 +1,38 @@
+source "subsys/net/Kconfig.template.log_config.net"
+
+config CAN_NET_NAME
+	string "Network interface name"
```
Isn't this the device name as typically our network interfaces do not have a name.

alexanderwachter replied:
Changed to `Network device name`

jukkar commented on include/net/can.h:
```diff
@@ -0,0 +1,186 @@
+ * +++   +---+   +----+--+   +-------+   +----+--+   +--+                     /
+ *  |    |                                               +-------------------+
+ *  +----+
+ */
```
I like the image :+1:

jukkar commented on include/net/can.h:
```diff
@@ -0,0 +1,186 @@
+ * Context for canbus net device.
+ */
+struct canbus_net_ctx {
+	int dad_filter_id;
```
Please document these. Also it might be better to put dad_work struct first and then others (could avoid holes although I am not claiming there are holes here).

jukkar commented on include/net/can.h:
```diff
@@ -0,0 +1,186 @@
+ * carry the rest of the message.
+ */
+struct can_net_isotp_tx_ctx {
+	/** Frame Control options received from FC frame */
```
Please order the fields in order to avoid holes in the struct. You can use pahole program to check that.

alexanderwachter replied:
Done. Thanks for pointing out pahole. I like this tool :-)

jukkar commented on include/net/can.h:
```diff
@@ -0,0 +1,186 @@
+
+
+struct can_net_isotp_rx_ctx {
+	/** Remaining data to receive. Goes from message length to zero */
```
same here

alexanderwachter replied:
4 bit holes Can't be avoided i guess ;-)

jukkar replied:
Yep :) But typically we can save bytes for each struct by carefully placing the fields in proper order. Not sure if this an issue here, it just looked there might be some optimization possibilities.

alexanderwachter replied:
I changed the order. pahole does not show XXX any more

jukkar commented on include/net/net_l2.h:
```diff
@@ -104,9 +104,15 @@ NET_L2_DECLARE_PUBLIC(BLUETOOTH_L2);
 #ifdef CONFIG_NET_L2_CANBUS
 #define CANBUS_L2		CANBUS
-#define CANBUS_L2_CTX_TYPE	void*
+#define CANBUS_L2_CTX_TYPE	struct net_can_context *
```
I recommend putting `#define CANBUS_L2_CTX_TYPE	struct net_can_context *` into canbus related header file.

alexanderwachter replied:
Isn't net_l2.h the right point for that? All other L2 CTX types are defined here. I don't like to include internal headers in other files using this definitions.

jukkar replied:
All the L2 CTX types that are defined in net_l2.h have a type "void *". But all the other L2 technologies that are using a real type, have their L2 CTX defined in relevant header file (for example ethernet.h and new ppp support that is coming). The benefit of this (to put the real context type into somewhere else) is that if you enable CONFIG_NET_L2_CANBUS, you need to include can.h in certain order so that compile will not complain about missing "struct net_can_context".

alexanderwachter replied:
Moved it to include/net/can.h

jukkar commented on subsys/net/l2/canbus/Kconfig:
```diff
@@ -0,0 +1,40 @@
+
+config NET_L2_CANBUS
+	bool "Enable CANBUS L2 layer [Experimental]"
```
EXPERIMENTAL here

jukkar commented on subsys/net/l2/canbus/Kconfig:
```diff
@@ -0,0 +1,40 @@
+	  Use a fixed L2 address for 6LoCAN instead of a random chosen one
+
+config NET_L2_CANBUS_FIXED_ADDR
+	int "L2 address"
```
Just pondering this but would it make sense to use hex value here, then it is easier to correlate the value if looking through message dumps etc.

jukkar commented on subsys/net/l2/canbus/canbus.c:
```diff
@@ -1,49 +1,1211 @@
+#define NET_CAN_DAD_SEND_RETRY 5
+#define NET_CAN_DAD_TIMEOUT K_MSEC(100)
+
+static struct can_net_isotp_tx_ctx tx_ctx[CONFIG_NET_PKT_TX_COUNT];
```
Could you create one context struct and then place these local variables there? We typically do it like that in the other L2 drivers as it is easier to read and review the code

jukkar commented on subsys/net/l2/canbus/canbus.c:
```diff
@@ -1,49 +1,1211 @@
+	net_pkt_lladdr_src(dst)->type = net_pkt_lladdr_src(src)->type;
+	net_buf_add_mem(buf, net_pkt_lladdr_src(src)->addr,
+			net_pkt_lladdr_src(src)->len);
+	net_buf_pull(buf, net_pkt_lladdr_src(src)->len);
```
Same comment to here for the net_buf_pull(), to me it looks like a wrong thing to do.

jukkar commented on subsys/net/l2/canbus/canbus.c:
```diff
@@ -1,49 +1,1211 @@
+	NET_ASSERT(net_lladdr->len == sizeof(u16_t));
+	return sys_be16_to_cpu(UNALIGNED_GET((u16_t *)net_lladdr->addr));
+}
+
```
remove extra empty line

jukkar commented on subsys/net/l2/canbus/canbus.c:
```diff
@@ -1,49 +1,1211 @@
+	if (!len) {
+		ret = net_pkt_read_be32(pkt, &len);
+	}
+	if (ret < 0) {
```
empty line after block ending }

jukkar commented on subsys/net/l2/canbus/canbus.c:
```diff
@@ -1,49 +1,1211 @@
+	frame->ext_id = addr_to_id(dest->addr, src->addr);
+}
+
+
```
extra empty line

jukkar commented on subsys/net/l2/canbus/canbus.c:
```diff
@@ -1,49 +1,1211 @@
+	set_frame_datalength(&frame, 3);
+
+	NET_DBG("Sending FC to ID: 0x%08x", frame.ext_id);
+	ret = api->send(net_can_dev, &frame, fc_send_cb, NULL, K_FOREVER);
```
I think you could just do `return api->send(net_can_dev, &frame, fc_send_cb, NULL, K_FOREVER);` and avoid having the ret variable.

jukkar commented on subsys/net/l2/canbus/canbus.c:
```diff
@@ -1,49 +1,1211 @@
+	src.addr = get_src_lladdr(pkt);
+	dest.addr = get_dest_lladdr(pkt);
+
+
```
remove extra empty line

jukkar commented on subsys/net/l2/canbus/canbus.c:
```diff
@@ -1,49 +1,1211 @@
+			return NET_OK;
+		}
+	}
+	/* At this point we expect to get Consecutive frames directly */
```
empty line after }

jukkar reviewed (changes requested):
Very nice PR :+1: 
Lot of code to review and did not managed to review it all so I will continue later.

jukkar commented on include/net/can.h:
```diff
@@ -0,0 +1,192 @@
+ * 250-   : Reserved
+ */
+struct can_net_fc_opts {
+	u8_t bs;     /** Block size. Number of CF PDUs before next CF is sent */
```
I think doxygen requires that comments are like this
`/**< Block size. Number of CF PDUs before next CF is sent */`
if the comment refers to things in the same line.

jukkar commented on drivers/can/can_net.c:
```diff
@@ -0,0 +1,254 @@
+
+static struct net_can_context net_can_context_1;
+
+NET_DEVICE_INIT(net_can_stm32_1, CONFIG_CAN_NET_NAME, net_can_init,
```
The name of the device  (1st parameter) looks wrong as it contains stm32, what about just calling it "net_can_1"?

alexanderwachter replied:
Copy n paste error.

jukkar commented on subsys/net/ip/ipv6_nbr.c:
```diff
@@ -976,6 +976,8 @@ static inline u8_t get_llao_len(struct net_if *iface)
 		return 8;
 	} else if (net_if_get_link_addr(iface)->len == 8U) {
 		return 16;
+	} else if (net_if_get_link_addr(iface)->len == 2) {
```
Perhaps 2U in order to follow the way the other values are set in this func.

jukkar commented on subsys/net/l2/canbus/canbus.c:
```diff
@@ -1,49 +1,1237 @@
+
+	rx_ctx->state = NET_CAN_RX_STATE_CF;
+	if (mcast) {
+		NET_INFO("Processed FF (multicast). Msg length: %u CTX: %p",
```
Not sure if it is useful to use the INFO level, it usually just prints too much data. Would NET_DBG() be more proper here?

alexanderwachter replied:
This was for debugging only. Changed it to DBG

jukkar commented on subsys/net/l2/canbus/canbus.c:
```diff
@@ -1,49 +1,1237 @@
+		NET_INFO("Processed FF (multicast). Msg length: %u CTX: %p",
+			msg_len, rx_ctx);
+	} else {
+		NET_INFO("Processed FF (unicast). Msg length: %u CTX: %p",
```
and here

jukkar commented on subsys/net/l2/canbus/canbus.c:
```diff
@@ -1,49 +1,1237 @@
+	if (rx_ctx) {
+		free_rx_ctx(rx_ctx);
+	}
+	return NET_DROP;
```
empty line bere return

jukkar commented on subsys/net/l2/canbus/canbus.c:
```diff
@@ -1,49 +1,1237 @@
+			tx_report_err(pkt);
+			return NET_DROP;
+		}
+		ctx->state = NET_CAN_TX_STATE_SEND_CF;
```
empty line after }

jukkar commented on subsys/net/l2/canbus/canbus.c:
```diff
@@ -1,49 +1,1237 @@
+	return ret;
+}
+
+
```
extra empty line

jukkar commented on subsys/net/l2/canbus/canbus.c:
```diff
@@ -1,49 +1,1237 @@
+	if (!ret) {
+		NET_DBG("Iface %p is up", iface);
+	}
+	return ret;
```
empty line before treturn

Commit:
net: l2: canbus: Add support for pkt reception from translator

This commits adds support for reception of packets thats comes from
a Ethernet to 6LoCAN translator. This packets carry the Ethernet
MAC address (6 bytes) inline in the FF (First Frame).

Signed-off-by: Alexander Wachter <alexander.wachter@student.tugraz.at>

Commit:
net: l2: canbus: Add support for canbus Ethernet translator

This commit adds support for a 6LoCAN Ethernet border translator.
CAN frames with the translator CAN address are translated and forwarded
to Ethernet. Ethernet frames with the first 34 bits matching the MAC
address of the translator are translated and forwarded to 6LoCAN.

Signed-off-by: Alexander Wachter <alexander.wachter@student.tugraz.at>

Commit:
net: ip: net_pkt: Implement net_pkt_shallow_clone

This commit implements net_pkt_shallow_clone. A shallow clone clones
the net_pkt but not the buffers. The buffers are only referenced and
therefor only freed when both copies of the net_pkt are freed.

Signed-off-by: Alexander Wachter <alexander.wachter@student.tugraz.at>

alexanderwachter commented:
6LoCAN now supports translation to Ethernet.
ndp has some problems but this will be fixed (need to translate ns and extend the source link-layer address option with the Ethernet MAC)

jukkar commented on subsys/net/l2/canbus/canbus.c:
```diff
@@ -175,6 +192,23 @@ static bool dest_is_translator(struct net_pkt *pkt)
+		return sizeof(struct net_eth_addr) +
+		       sizeof(struct net_can_lladdr);
+	}
+#else
```
This looks ugly, can't we use same function prototype for total_lladdr_len()?
So something like:
```
static size_t total_lladdr_len(struct net_pkt *pkt)
{
#if defined(CONFIG_NET_L2_CANBUS_ETH_TRANSLATOR)
	/* This pkt will be farowarded to Ethernet
	 * Destination MAC is carried inline, source is going to be extended
	 */
	if (dest_is_translator(pkt)) {
		return sizeof(struct net_eth_addr) +
		       sizeof(struct net_can_lladdr);
	}
#else 
	ARG_UNUSED(pkt);
#endif 

	return 2U * sizeof(struct net_can_lladdr);
}
```

alexanderwachter replied:
Sure

jukkar commented on drivers/can/can_net.c:
```diff
@@ -183,6 +186,33 @@ static inline int attach_mcast_filter(struct net_can_context *ctx)
 	return filter_id;
 }
 
+#ifdef CONFIG_NET_L2_CANBUS_ETH_TRANSLATOR
```
Minor nitpick but we are preferring `#if defined(xxx)` in networking code, Could use change all of the `#ifdef` to those?

jukkar commented on drivers/can/can_net.c:
```diff
@@ -183,6 +186,33 @@ static inline int attach_mcast_filter(struct net_can_context *ctx)
+		return CAN_NET_FILTER_NOT_SET;
+	}
+
+	NET_DBG("Attached ETH bridge filter %d.", filter_id);
```
No need for `.` at the end of the string (saves one byte here).

jukkar commented on drivers/can/can_net.c:
```diff
@@ -183,6 +186,33 @@ static inline int attach_mcast_filter(struct net_can_context *ctx)
+
+	return filter_id;
+}
+#endif /*CONFIG_NET_L2_CANBUS_ETH_TRANSLATOR*/
```
Another minor nitpick. Instead of `/*xxx*/`, could you write these comments with spaces after and before comment, so `/* xxx */`

jukkar commented on subsys/net/l2/canbus/canbus.c:
```diff
@@ -186,6 +220,13 @@ static inline void cpy_lladdr(struct net_pkt *dst, struct net_pkt *src)
 
+#ifdef CONFIG_NET_L2_CANBUS_ETH_TRANSLATOR
+	if (dest_is_translator(src)) {
+		/*Make room for address extension*/
```
Please add spaces here `/* Make room .... extension */`
Can you check all the comments for these and fix them all.

jukkar commented on subsys/net/l2/canbus/canbus.c:
```diff
@@ -809,6 +867,18 @@ static inline int send_ff(struct net_pkt *pkt, size_t len, bool mcast)
+		   (NET_CAN_DL > (6 + net_pkt_lladdr_src(pkt)->len)) ||
+		   len <= 0x0FFF);
+	NET_ASSERT(mcast || !(dest_is_translator(pkt) && src_is_translator(pkt)));
+	if (src_is_translator(pkt)) {
```
Empty line before the if() would look nicer.

jukkar commented on subsys/net/l2/canbus/canbus.c:
```diff
@@ -1002,6 +1082,191 @@ static enum net_verdict process_frame(struct net_pkt *pkt)
+{
+	struct net_if *canbus_iface;
+
+	canbus_iface = net_if_get_first_by_type(&NET_L2_GET_NAME(CANBUS));
```
Initially this is ok, but we need to figure out what to if there are multiple CAN interfaces in the system.

jukkar commented on subsys/net/l2/canbus/canbus.c:
```diff
@@ -1002,6 +1082,191 @@ static enum net_verdict process_frame(struct net_pkt *pkt)
+	net_if_queue_tx(canbus_iface, pkt);
+}
+
+static struct net_ipv6_hdr *get_ip_hdr_from_eth_frame(struct net_pkt *pkt)
```
There already exists `NET_IPV6_HDR(pkt)` macro that will return this so this function should not be needed.

alexanderwachter replied:
I have to skip the Ethernet heade first.

jukkar commented on subsys/net/l2/canbus/canbus.c:
```diff
@@ -1002,6 +1082,191 @@ static enum net_verdict process_frame(struct net_pkt *pkt)
+
+	/*This frame is for the Ethernet interface itself*/
+	if (net_linkaddr_cmp(net_if_get_link_addr(iface), lladdr) ||
+	    net_eth_is_addr_lldp_multicast((struct net_eth_addr *)lladdr->addr)) {
```
Note sure I follow you here, what is the idea for checking lldp multicast address?

alexanderwachter replied:
LLDP is not supported by 6LoCAN

jukkar commented on subsys/net/l2/canbus/canbus.c:
```diff
@@ -1002,6 +1082,191 @@ static enum net_verdict process_frame(struct net_pkt *pkt)
+		return NET_CONTINUE;
+	}
+
+	eth_iface = net_if_get_first_by_type(&NET_L2_GET_NAME(ETHERNET));
```
Initially ok, but does not work if there are multiple ethernet interfaces (like in VLAN case).

alexanderwachter replied:
True. Maybe use a Kconfig for that?

jukkar replied:
Lets do it like this in version 1, and change it to support multiple interfaces in version 2. Otherwise there is just too many changes going around.

Commit:
net: ip: ipv6_nbr: take TLLAO length from src lladdres

Take the targer link layer address option link laxer addres lenght
from net packets source addres instead if the interface.
This is usefull for 6LoCAN border translator (6LoCAN to Ethernet).

Signed-off-by: Alexander Wachter <alexander.wachter@student.tugraz.at>

Commit:
net: ip: ipv6_nbr: use temp variables for ntoh conversion

The network to host byte order conversion is actually in place
on the network buffer. This prevents the reuse and forwarding of
RA packets. This commit uses temporary variables instead.

Signed-off-by: Alexander Wachter <alexander.wachter@student.tugraz.at>

jukkar commented on subsys/net/l2/canbus/canbus.c:
```diff
@@ -200,10 +200,28 @@ static bool dest_is_translator(struct net_pkt *pkt)
+	/* This pkt will be farowarded to Ethernet
+	 * Destination MAC is carried inline, source is going to be extended
+	 */
+	if (is_for_translator(pkt)) {
```
What about doing it like this instead:
```
if (IS_ENABLED(CONFIG_NET_L2_CANBUS_ETH_TRANSLATOR) && is_for_translator(pkt)) {
```
Then you could also remove the #ifdef's around `is_for_translator()`

jukkar commented on subsys/net/l2/canbus/canbus.c:
```diff
@@ -218,6 +236,13 @@ static inline void cpy_lladdr(struct net_pkt *dst, struct net_pkt *src)
 	lladdr->type = NET_LINK_CANBUS;
 
+#ifdef CONFIG_NET_L2_CANBUS_ETH_TRANSLATOR
+	if (is_for_translator(src)) {
```
same here

jukkar commented on subsys/net/l2/canbus/canbus.c:
```diff
@@ -347,6 +372,19 @@ static enum net_verdict finish_pkt(struct net_pkt *pkt)
 				  net_pkt_lladdr_src(pkt)->len);
 
+#if defined(CONFIG_NET_L2_CANBUS_ETH_TRANSLATOR)
+	if (is_for_translator(pkt)) {
```
and here too

jukkar commented on subsys/net/l2/canbus/canbus.c:
```diff
@@ -347,6 +372,19 @@ static enum net_verdict finish_pkt(struct net_pkt *pkt)
+		net_pkt_lladdr_dst(pkt)->addr = net_pkt_cursor_get_pos(pkt);
+		net_pkt_lladdr_dst(pkt)->type = NET_LINK_ETHERNET;
+		net_pkt_lladdr_dst(pkt)->len = sizeof(struct net_eth_addr);
+		net_buf_pull(pkt->buffer, sizeof(struct net_eth_addr));
```
Hmm, I think you should init the cursor *after* you have pulled stuff, otherwise the cursor would point to "unexpected" position (unless you want that of course).

jukkar commented on subsys/net/l2/canbus/canbus.c:
```diff
@@ -886,6 +938,14 @@ static inline int send_ff(struct net_pkt *pkt, size_t len, bool mcast,
 	}
 
+#ifdef CONFIG_NET_L2_CANBUS_ETH_TRANSLATOR
+	if (net_pkt_lladdr_src(pkt)->type == NET_LINK_ETHERNET) {
```
Please remove the ifdef and put IS_ENABLED(CONFIG_NET_L2_CANBUS_ETH_TRANSLATOR) info if statement.

jukkar commented on subsys/net/l2/canbus/canbus.c:
```diff
@@ -1077,6 +1148,329 @@ static enum net_verdict process_frame(struct net_pkt *pkt)
+	struct net_pkt *clone_pkt;
+	struct net_if *canbus_iface;
+
+	/*Forward only IPv6 frames*/
```
Could you add spaces here (and also the comments in other parts of this PR). So `/* xxx */` instead of `/*xxx*/`

jukkar commented on subsys/net/l2/canbus/canbus.c:
```diff
@@ -1096,9 +1490,14 @@ static enum net_verdict canbus_recv(struct net_if *iface,
 		if (pkt->can_rx_ctx->state == NET_CAN_RX_STATE_FIN) {
 			rx_finish(pkt);
+
+#ifdef CONFIG_NET_L2_CANBUS_ETH_TRANSLATOR
```
Change this to use IS_ENABLED(CONFIG_NET_L2_CANBUS_ETH_TRANSLATOR) instead so that we get the compiler to check both branches.

jukkar commented on subsys/net/ip/net_pkt.c:
```diff
@@ -1868,6 +1868,8 @@ bool net_pkt_is_contiguous(struct net_pkt *pkt, size_t size)
 	if (pkt->cursor.buf && pkt->cursor.pos) {
```
There is some extra characters after the signed-off-line in commit "net: ip: net_pkt: Add pkt_cursor_advance before contiguous check"

alexanderwachter replied:
This Commit is not part of this PR. Original PR: 16201

jukkar reviewed (changes requested):
Looks good, just some minor tweaks need and we are good to go.

alexanderwachter commented:
Followin commits are not part of this PR:
```
f9d508c73c net: ip: 6lo compression rework
fe3febc670 net: ip: net_pkt: Add pkt_cursor_advance before contiguous check
5b84c67de0 net: ip: 6lo: remove memmove and net_compact from IPHC uncompress
779bb8e14e net: ip: 6lo: Remove unnecessary net_buf allocation and memmove
2f411bb98e drivers: can: stm32: Remove STM CAN_Init
e1050ce417 drivers: can: stm32: Change TX priority to chronological order
```

alexanderwachter commented on drivers/can/can_net.c:
```diff
@@ -159,29 +256,56 @@ static inline int attach_unicast_filter(struct net_can_context *ctx)
+static inline int attach_eth_bridge_filter(struct net_can_context *ctx)
 {
-	struct zcan_filter filter = {
+	static struct zcan_filter filter = {
```
This should be const.

alexanderwachter commented on drivers/can/can_net.c:
```diff
@@ -159,29 +256,56 @@ static inline int attach_unicast_filter(struct net_can_context *ctx)
+
+static inline int attach_all_mcast_filter(struct net_can_context *ctx)
+{
+	static struct zcan_filter filter = {
```
Done.

Commit:
samples: net: echo_client and echo_server: Add 6LoCAN config

This commit adds a sample configuration for 6LoCAN to the
echo_client and echo_server example.

Signed-off-by: Alexander Wachter <alexander.wachter@student.tugraz.at>

jukkar reviewed (approved):
LGTM

henrikbrixandersen commented:
@alexanderwachter Looks good. I have looked allover for other implementations and/or references to "6LoCAN" but I have not found any?

alexanderwachter commented:
@henrikbrixandersen Thanks for reviewing!
>I have looked allover for other implementations and/or references to "6LoCAN" but I have not found any?
This PR is the very first implementation of 6LoCAN. I'm currently writing an IETF-draft for it. I'll post the draft when it's ready.

alexanderwachter commented:
@jukkar There are still some minor changes. I made some simplifications in the standard draft. Hope they are ready at the end of the week.

pfalcon commented:
@alexanderwachter : I guess it would be helpful for reviewers to know whether you target this for the next release of Zephyr (2.0 or whatever) or not (code freeze a week after next, or something).

alexanderwachter commented:
@pfalcon didn't think about it yet. Maybe we should wait until I have released the IETF-draft, but actually, I don't reckon mayor changes. Merging it in 2.0 would help because I could reference it as the reference implementation.

alexanderwachter commented:
Rebased and made some simplifications. Dropped support for"Jumbo frames" and always use 8-bit data length for single frames.

alexanderwachter commented:
@pfalcon here is the plan: we want to merge it into 2.0.
Can you review?

tbursztyka commented on include/net/net_pkt.h:
```diff
@@ -1525,6 +1525,16 @@ int net_pkt_copy(struct net_pkt *pkt_dst,
+ * @brief Clone pkt and increase the refcount of its buffer.
+ *
+ * @param pkt Original pkt to be shallow cloned
+ * @param timeout Timeout to wait for free buffer
```
"for free packet" the buffer are just refcounted and not allocated

alexanderwachter replied:
True! Updated comment.

tbursztyka commented on subsys/net/ip/net_pkt.c:
```diff
@@ -1719,6 +1724,46 @@ struct net_pkt *net_pkt_clone(struct net_pkt *pkt, s32_t timeout)
+	clone_pkt->buffer = pkt->buffer;
+
+	buf = pkt->buffer;
+
```
remove that empty line

tbursztyka commented on drivers/can/can_net.c:
```diff
@@ -12,13 +12,39 @@ LOG_MODULE_REGISTER(net_can, CONFIG_CAN_NET_LOG_LEVEL);
 
```
revise the function names in this commit. Public functions should always have net_<domain>_ prefix (so here: net_can_).

And it is nicer to have can_ as prefix for local function. For instance is_for_translator() -> can_is_pkt_for_translator()

Apply that everywhere

alexanderwachter replied:
Done.

tbursztyka commented on drivers/can/can_net.c:
```diff
@@ -0,0 +1,254 @@
+
+drop:
+	NET_INFO("pkt dropped");
+	if (pkt) {
```
have an empty line befor the if

tbursztyka commented on subsys/net/l2/canbus/canbus.c:
```diff
@@ -1,49 +1,1333 @@
 /*
- * Copyright (c) 2019 Intel Corporation.
+ * Copyright (c) 2019 Alexander Wachter.
```
No, let the original ©, just add yours below.

pfalcon replied:
> No, let the original ©, just add yours below.

Umm, no, separate renaming of the old CANBUS to CANBUS_RAW vs creation of new CANBUS into 2 separate commits, so it was clear what happens here. And perhaps, it should be canbus/6locan.c after all...

alexanderwachter replied:
@tbursztyka This is because I changed the file-names. The canbus.c is (c) by, but the canbus_raw.c is (c) by Intel. I'll change it to 6locan as @pfalcon suggested.

tbursztyka replied:
ok missed that

tbursztyka commented on subsys/net/l2/canbus/canbus.c:
```diff
@@ -1,49 +1,1333 @@
+char *net_sprint_addr(sa_family_t af, const void *addr);
+
+#if CONFIG_NET_L2_CANBUS_LOG_LEVEL >= LOG_LEVEL_DBG
+static void print_ip_hdr(struct net_ipv6_hdr *ip_hdr)
```
For all local function, add canbus_ prefix

alexanderwachter replied:
Done.

tbursztyka commented on subsys/net/l2/canbus/canbus.c:
```diff
@@ -1,49 +1,1333 @@
+				dest.addr = get_dest_lladdr(pkt);
+				net_can_dev = net_if_get_device(pkt->iface);
+				ret = send_fc(net_can_dev, &src, &dest,
+					NET_CAN_PCI_FS_CTS);
```
indentation

tbursztyka reviewed (changes requested):
tiny changes

pfalcon commented:
Commit message:

> This commit is an implementation of 6LoCAN, a 6Lo adaption layer for
Controller Area Networks. 6LoCAN is not yet standardised.

If there's a draft RFC, I'd suggest to link to it.

pfalcon commented on subsys/net/l2/canbus/Kconfig:
```diff
@@ -0,0 +1,68 @@
+	  Greater than one only makes sense for random link layer addresses.
+
+config NET_L2_CANBUS_STMIN
+	int "STmin"
```
I'd suggest to definitely use more detailed description, and perhaps use more detailed option name.

pfalcon reviewed (changes requested):
> @pfalcon here is the plan: we want to merge it into 2.0.
Can you review?

Yes, I can. And there're definitely things to clean up here... (looked at just 1 commit so far).

alexanderwachter commented on drivers/can/Kconfig.net:
```diff
@@ -0,0 +1,37 @@
+	default 80
+	help
+	  CAN NET device driver initialization priority.
+	  Do not mess with it unless you know what you are doing.
```
@pfalcon This is copied from ethernet.
With `grep -R "Do not mess with it unless you know what you are doing."  | wc -l` you get a result of 5.

pfalcon replied:
Ok then.

Just to clarify, subject of -1 was an obvious define duplication, then comment syntax. Everything else are for your consideration.

alexanderwachter replied:
Trying to address all of that. I appreciate all kind of comments.

alexanderwachter commented on drivers/can/can_net.c:
```diff
@@ -0,0 +1,254 @@
+{
+	struct net_buf *buf = pkt->buffer;
+
+	net_pkt_lladdr_dst(pkt)->addr = buf->data;
```
>You receive empty pkt?

This pkt was allocated before and is empty, yes.

>The you try to store can_lladdr's at the beginning of it, point other pkt lladdr fields inside its own buffer?

That is correct. Other l2 do the same. ETH, for example, sets this the ptr to the address in the ETH header and then pull it. The problem is that there is no other memmory reserved for lladdr in the pkt. I guess it's because of the different sizes.

>You use different values for the sizes of these field, like when it's referred as and address, it's sizeof(struct net_can_lladdr), while when you put it into buffer, it's sizeof(u16_t)? Do you expect them to match?

`sizeof(struct net_can_lladdr)` is actually 2 octets. I need to change the byte order to net-byte-order, so I have to use u16_t as type.

pfalcon replied:
> Other l2 do the same.

Ok. (I assume they do it the same with the same, uncommented code.)

> sizeof(struct net_can_lladdr) is actually 2 octets.

Yeah, I figured. What's somewhat confusing is why in one case you use `sizeof(struct net_can_lladdr)` and in other `sizeof(u16_t)`.

> I need to change the byte order to net-byte-order, so I have to use u16_t as type.

Noting that, it indeed makes sense to use different types there, though it still looks a bit weird at first sight. I'd personally looked to add a code comment if I wrote such code (not sure if I'd succeed).

All in all, that's relatively minor point comparing with the rest of code.

alexanderwachter commented on include/net/can.h:
```diff
@@ -0,0 +1,215 @@
+	/** The interface associated with this device */
+	struct net_if *iface;
+	/** The link layer address chosen for this interface */
+	u16_t ll_addr;
```
Because Zephyr expects link-layer address as a u8_t pointer. The upper two bits are always zero. This address is the address when you call net_if_get_link_addr().

alexanderwachter commented on include/net/can.h:
```diff
@@ -0,0 +1,215 @@
+};
+
+/*
+ * STmin is split in two ranges:
```
Two valid ones. I'll update the comment.

alexanderwachter commented on include/net/can.h:
```diff
@@ -0,0 +1,215 @@
+ */
+struct can_net_fc_opts {
+	/** Block size. Number of CF PDUs before next CF is sent */
+	u8_t bs;
```
The term BS is coming from the ISO-TP standard which is unfortunately not available publicly.
I'll add comment block for all the abbreviations defined in the standard.

alexanderwachter commented on include/net/can.h:
```diff
@@ -0,0 +1,215 @@
+	/** Actual block number that is transmitted. Counts from BS to 0 */
+	u8_t act_block_nr;
+	/** Number of WAIT frames received*/
+	u8_t wft;
```
Abbreviation used by the ISO-TP standard. This "counter" is called N_WFT there.

alexanderwachter commented on drivers/can/can_net.c:
```diff
@@ -0,0 +1,254 @@
+{
+	struct net_can_context *ctx = (struct net_can_context *)arg;
+	size_t pkt_size = 2 * sizeof(struct net_can_lladdr) +
+			  get_frame_datalength(frame);
```
Prefix added.

alexanderwachter commented on include/net/can.h:
```diff
@@ -0,0 +1,215 @@
+
+/**
+ * Context for a transmission of messages that didn't fit in a single frame.
+ * This messages Start with a FF (First Frame) that is in case of unicast
```
Done.

alexanderwachter commented on include/net/can.h:
```diff
@@ -0,0 +1,215 @@
+	struct can_net_fc_opts opts;
+	/** CAN destination address */
+	struct net_can_lladdr dest_addr;
+	/** Remaining data to transbit in bytes */
```
Typo.

alexanderwachter commented on include/net/can.h:
```diff
@@ -0,0 +1,215 @@
+};
+
+
+struct can_net_isotp_rx_ctx {
```
Forgot to add a comment block.

Commit:
net: canbus: Rename canbus to canbus_raw

Rename the socket_can implementation from CANBUS to CANBUS_RAW.
This is a preperation for 6LoCAN which is a CANBUS L2 for IPv6.

Signed-off-by: Alexander Wachter <alexander.wachter@student.tugraz.at>

Commit:
net: l2: 6LoCAN implementation

This commit is an implementation of 6LoCAN, a 6Lo adaption layer for
Controller Area Networks. 6LoCAN is not yet standardised.

Signed-off-by: Alexander Wachter <alexander.wachter@student.tugraz.at>

jukkar commented:
@alexanderwachter Could you rebase against very latest master and resubmit, perhaps it solves the unrelated and failing sanitycheck.

alexanderwachter commented:
@jukkar thought Shippable always applies the patch on the latest master. What is the best way to restart the test? Close and reopen the PR?

alexanderwachter commented on include/net/net_pkt.h:
```diff
@@ -207,6 +209,12 @@ struct net_pkt {
+#endif
+#if defined(CONFIG_NET_L2_CANBUS)
+	union {
+		struct can_net_isotp_tx_ctx *can_tx_ctx;
```
Usually, you have only one of those technologies at the same time. An alternative would be to have a list and search those lists every time a new packet arrives.

alexanderwachter commented on samples/net/sockets/can/src/main.c:
```diff
@@ -170,7 +170,7 @@ static int setup_socket(void)
 	can_copy_zfilter_to_filter(&zfilter, &filter);
 
-	iface = net_if_get_first_by_type(&NET_L2_GET_NAME(CANBUS));
+	iface = net_if_get_first_by_type(&NET_L2_GET_NAME(CANBUS_RAW));
```
Done.

jukkar commented:
> thought Shippable always applies the patch on the latest master. What is the best way to restart the test? Close and reopen the PR?

Shippable uses whatever you send to it, iirc it does not rebase the tree. Currently closing->re-opening is the easiest way to restart the test.

alexanderwachter commented on subsys/net/ip/ipv6_nbr.c:
```diff
@@ -976,6 +976,8 @@ static inline u8_t get_llao_len(struct net_if *iface)
 		return 8;
 	} else if (net_if_get_link_addr(iface)->len == 8U) {
 		return 16;
+	} else if (net_if_get_link_addr(iface)->len == 2U) {
```
@jukkar @pfalcon @tbursztyka should we use a switch case or ROUND_UP(net_if_get_link_addr(iface)->len, 8) here?

jukkar replied:
ROUND_UP(8, 8) gives result of 8 instead of 16.

alexanderwachter replied:
@jukkar forgot the +2 for type and length.

jukkar replied:
I am fine with using ROUND_UP() here

alexanderwachter replied:
I'll do it in a separate PR.

jukkar commented:
ping @pfalcon @tbursztyka could you re-review

pfalcon reviewed (approved):
Thanks for updates, let's go with it.
